### PR TITLE
Correctly handle recoding pair with value range and source containing missings

### LINKED ISSUE #105: recode fails with missing and value ranges
This works:
```
julia> recode([1,2,3], [1,2]=>100)
3-element Array{Int64,1}:
 100
 100
   3
```
and this works:
```
julia> recode(categorical([1,2,3,missing]), [1,2]=>100)
4-element CategoricalArrays.CategoricalArray{Union{Int64, Missings.Missing},1,UInt32}:
 100
 100
 3
 missing
```
but this fails:
```
julia> recode([1,2,3,missing], [1,2]=>100)
ERROR: TypeError: non-boolean (Missings.Missing) used in boolean context
Stacktrace:
 [1] any(::Base.##136#137{Missings.Missing}, ::Array{Int64,1}) at .\reduce.jl:574
 [2] in(::Missings.Missing, ::Array{Int64,1}) at .\reduce.jl:631
 [3] recode!(::Array{Union{Int64, Missings.Missing},1}, ::Array{Union{Int64, Missings.Missing},1}, ::Void, ::Pair{Array{Int64,1},Int64}, ::Vararg{Pair{Array{Int64,1},Int64},N} where N) at D:\Software\JULIA_PKG\v0.6\CategoricalArrays\src\recode.jl:39
 [4] recode(::Array{Union{Int64, Missings.Missing},1}, ::Void, ::Pair{Array{Int64,1},Int64}, ::Vararg{Pair{Array{Int64,1},Int64},N} where N) at D:\Software\JULIA_PKG\v0.6\CategoricalArrays\src\recode.jl:332
 [5] recode(::Array{Union{Int64, Missings.Missing},1}, ::Pair{Array{Int64,1},Int64}) at D:\Software\JULIA_PKG\v0.6\CategoricalArrays\src\recode.jl:317
```

Probably line:
https://github.com/JuliaData/CategoricalArrays.jl/blob/master/src/recode.jl#L39
should be fixed.

### PR DESCRIPTION
**Pull request by bkamins**
An attempt to fix #105.
I have used `any(x ≅ y for y in p.first)` condition.

@nalimilan The tests are really mind blowing. I hope I have not mixed up anything.

I have also made a small addition to docstrings about non-obvious handling of levels ordering when using `default` in `recode` that came up during tests.

### CODE REVIEW

**Comment by bkamins** [comment]
https://github.com/JuliaData/CategoricalArrays.jl/pull/106/files#diff-f538847e613490e5199ee41f46c32799R181
fixes another bug (when `missing` is `first` in recode pair, but is hidden inside a collection).

**Comment by bkamins** [comment]
@nalimilan Tests fail on Julia latest, but it seems unrelated. Yes?

File: src/recode.jl [hunk 1]
```diff
@@ -11,6 +11,8 @@ or [`in`](@ref) the key (first item of the pair), then the corresponding value
 (second item) is copied to `dest`.
 If the element matches no key and `default` is not provided or `nothing`, it is copied as-is;
 if `default` is specified, it is used in place of the original element.
+If `dest` is CategoricalArray and `default` is provided then the ordering of resulting levels
```
Comment (nalimilan) [inline src/recode.jl]: Good idea. But the order of pairs is always used for the levels IIRC, it's just that when `default` is not provided original levels appear first. I guess you can just omit "and `default` is provided" but say "`default` will be the last level if provided".

Also use backticks around `CategoricalArray`, and I'd say "a `CategoricalArray`. Finally, I think it would make sense to mention this only for a `CategoricalArray` method, after showing its signature, just like it's currently done for arrays supporting missing values.

File: src/recode.jl [hunk 2]
```diff
@@ -11,6 +11,8 @@ or [`in`](@ref) the key (first item of the pair), then the corresponding value
 (second item) is copied to `dest`.
```
Comment (nalimilan) [inline src/recode.jl]: On the previous line, mentioning `in` is no longer correct. Maybe say something like ``if the element is equal to (according to [`isequal`](@ref))) the key (first item of the pair) or to one of its entries if it is a collection``.

File: src/recode.jl [hunk 3]
```diff
@@ -176,7 +178,7 @@ function recode!(dest::CategoricalArray{T}, src::CategoricalArray, default::Any,
     # For missing values (0 if no missing in pairs' keys)
     indexmap[1] = 0
     for p in pairs
-        if ismissing(p.first)
+        if any(ismissing.(p.first))
```
Comment (nalimilan) [inline src/recode.jl]: Better use `any(ismissing, p.first)`.

File: test/16_recode.jl [hunk 4]
```diff
@@ -159,7 +159,21 @@ end
     end
 end
 
-@testset "Recoding array with missings, no default and with missing as a key pair from $(typeof(x)) to $(typeof(y))" for
+@testset "Group recoding array with missings and no default from $(typeof(x)) to $(typeof(y))" for
```
Comment (nalimilan) [inline test/16_recode.jl]: Rather than "group recoding", I'd find it more explicit to say "collection in LHS" or "collection with missing in LHS" depending on the case.

File: test/16_recode.jl [hunk 5]
```diff
@@ -173,6 +187,20 @@ end
     end
 end
 
+@testset "Group recoding array with missings, default and with missing as a key pair from $(typeof(x)) to $(typeof(y))" for
```
Comment (nalimilan) [inline test/16_recode.jl]: Mention in the test name that `missing` is the first entry in a collection LHS.

File: src/recode.jl (hunk 2 above)
Comment (bkamins) [reply src/recode.jl]: fixed

File: src/recode.jl (hunk 1 above)
Comment (bkamins) [reply src/recode.jl]: fixed

File: src/recode.jl (hunk 3 above)
Comment (bkamins) [reply src/recode.jl]: agreed, fixed

File: test/16_recode.jl (hunk 4 above)
Comment (bkamins) [reply test/16_recode.jl]: fixed

File: test/16_recode.jl (hunk 5 above)
Comment (bkamins) [reply test/16_recode.jl]: fixed

File: src/recode.jl (hunk 3 above)
Comment (bkamins) [reply src/recode.jl]: @nalimilan Now I remember why I did not used your code :). In the last commit I have rewritten all tests of `p.fits` in a cleaner and consistent manner handling both vectors and atoms in `p.first`.

**Comment by bkamins** [comment]
@nalimilan As a general comment: we have a corner case if a variable is used to store `Union{Tuple, AbstractArray}`, e.g.:
```
julia> x = [[1,2], [2,3], [3,4]]
3-element Array{Array{Int64,1},1}:
 [1, 2]
 [2, 3]
 [3, 4]

julia> recode(x, [1,2] => [100,101])
3-element Array{Array{Int64,1},1}:
 [1, 2]
 [2, 3]
 [3, 4]

julia> recode(x, [[1,2]] => [100,101])
3-element Array{Array{Int64,1},1}:
 [100, 101]
 [2, 3]
 [3, 4]
```

Not sure if you want to include a comment in the docs that if you have such a situation you have to wrap LHS in a vector or tulpe?

File: test/16_recode.jl [hunk 6]
```diff
@@ -159,7 +159,21 @@ end
     end
 end
 
-@testset "Recoding array with missings, no default and with missing as a key pair from $(typeof(x)) to $(typeof(y))" for
+@testset "Collection in LHS recoding array with missings and no default from $(typeof(x)) to $(typeof(y))" for
+    x in (["1", missing, "3", "4", "5"], CategoricalArray(["1", missing, "3", "4", "5"])),
+    y in (similar(x), Array{Union{String, Missing}}(size(x)),
+          CategoricalArray{Union{String, Missing}}(size(x)), x)
+
```
Comment (nalimilan) [inline test/16_recode.jl]: No empty line here either, nor below.

File: test/16_recode.jl [hunk 7]
```diff
@@ -173,6 +187,20 @@ end
     end
 end
 
+@testset "Collection with missing in LHS recoding array with missings, default and with missing as a key pair from $(typeof(x)) to $(typeof(y))" for
```
Comment (nalimilan) [inline test/16_recode.jl]: "and with missing as a key pair" isn't correct. Same below.

File: src/recode.jl [hunk 8]
```diff
@@ -36,8 +41,9 @@ function recode!(dest::AbstractArray{T}, src::AbstractArray, default::Any, pairs
 
         for j in 1:length(pairs)
             p = pairs[j]
-            if (!isa(p.first, Union{AbstractArray, Tuple}) && x ≅ p.first) ||
-               (isa(p.first, Union{AbstractArray, Tuple}) && x in p.first)
+            if ((isa(p.first, Union{AbstractArray, Tuple}) && any(x ≅ y for y in p.first)) ||
+                x ≅ p.first)
```
[the remark this answers is not in the document]
Comment (bkamins) [reply src/recode.jl]: blanks removed.
Regarding the code style I have compared:

```
function f1(x, z)
    if (isa(z, Union{AbstractArray, Tuple}) && any(isequal(x, y) for y in z)) || isequal(x, z)
       return 1
    end
    return 0
end

function f2(x, z)
    if (!isa(z, Union{AbstractArray, Tuple}) && isequal(x, z)) ||
       (isa(z, Union{AbstractArray, Tuple}) && any(isequal(x, y) for y in z))
       return 1
    end
    return 0
end
```

on Julia 0.7 and `@code_llvm` produces on them essentially identical code. I can revert the original code, but I felt it was harder to read.

File: test/16_recode.jl (hunk 6 above)
Comment (bkamins) [reply test/16_recode.jl]: fixed those empty lines everywhere in the file. Please have a look if this is what you have intended.

File: test/16_recode.jl (hunk 7 above)
Comment (bkamins) [reply test/16_recode.jl]: fixed

File: src/recode.jl [hunk 9]
```diff
@@ -36,8 +41,8 @@ function recode!(dest::AbstractArray{T}, src::AbstractArray, default::Any, pairs
...
```
Comment (nalimilan) [inline src/recode.jl]: Trailing whitespace. Same below.

File: src/recode.jl [hunk 10]
```diff
@@ -189,8 +195,9 @@ function recode!(dest::CategoricalArray{T}, src::CategoricalArray, default::Any,
     @inbounds for (i, l) in enumerate(srcindex)
         for j in 1:length(pairs)
             p = pairs[j]
-            if (!isa(p.first, Union{AbstractArray, Tuple}) && l ≅ p.first) ||
-               (isa(p.first, Union{AbstractArray, Tuple}) && l in p.first)
+            if ((isa(p.first, Union{AbstractArray, Tuple}) && any(l ≅ y for y in p.first)) ||
+                l ≅ p.first) 
+
```
Comment (nalimilan) [inline src/recode.jl]: Remove empty line.

File: test/16_recode.jl [hunk 11]
```diff
@@ -14,7 +14,6 @@ const ≅ = isequal
     x in ([1:10;], CategoricalArray(1:10), CategoricalArray{Union{Int, Missing}}(1:10)),
     y in (similar(x), Array{Int}(size(x)),
           CategoricalArray{Int}(size(x)), CategoricalArray{Union{Int, Missing}}(size(x)), x)
-
```
Comment (nalimilan) [inline test/16_recode.jl]: OK, the package doesn't appear to be using a completely consistent style. In general PRs should adopt whatever style is used in a given file, with the goal of keeping them as consistent as possible and limiting the diff size. Here, better not remove these empty lines, and use the same style below (but not in the other file).

File: src/recode.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (nalimilan) [reply src/recode.jl]: Good. Though just to be sure, please confirm your tested passing both scalars and tuples/arrays.

**Comment by nalimilan** [comment]
> Not sure if you want to include a comment in the docs that if you have such a situation you have to wrap LHS in a vector or tulpe?

To be honest `recode` is really not intended at this kind of situation. After all, it lives in a package which isn't intended at storing arrays inside arrays either.

File: src/recode.jl (hunk 9 above)
Comment (bkamins) [reply src/recode.jl]: fixed

File: src/recode.jl (hunk 10 above)
Comment (bkamins) [reply src/recode.jl]: fixed

File: test/16_recode.jl (hunk 11 above)
Comment (bkamins) [reply test/16_recode.jl]: I have put a newline after every multiline for

**Comment by bkamins** [comment]
Yes - tests cover LHS:
* `missing`
* non-missing
* `missing` in vector
* vector of only non-missings

**Comment by nalimilan** [comment]
Actually, GitHub lost track of the parent comment, but by tests I referred to `@code_llvm` for the `isa` branches, not to `@test`.

**Comment by bkamins** [comment]
Checked both - seem the same. I am not clear why they might be different, we had:
`(a && b) || (!a && c)` which I assume for any reasonable compiler should be equivalent to `(a && b) || c`.

Can you explain why it might not be the case (those are areas of Julia I am not very familiar with)?

**Comment by nalimilan** [comment]
I don't know, I just wanted to be sure, because compiler optimizations can be brittle sometimes.
